PyInstaller/utils/osx.py

```diff
@@ -80,29 +84,208 @@ def fix_exe_for_code_signing(filename):
+    """
+    executable = MachO(filename)
+    return bool(executable.fat), [_get_arch_string(hdr.header)
+                                  for hdr in executable.headers]
+
+
+def convert_binary_to_thin_arch(filename, thin_arch):
+    """
+    Convert the given fat binary into thin one with the specified
+    target architecture.
+    """
+    cmd_args = ['lipo', '-thin', thin_arch, filename, '-output', filename]
+    retcode, stdout, stderr = exec_command_all(*cmd_args)
+    if retcode != 0:
+        logger.warning("lipo command (%r) failed with error code %d!\n"
+                       "stdout: %r\n"
+                       "stderr: %r",
+                       cmd_args, retcode, stdout, stderr)
+        raise SystemError("lipo failure!")
+
+
+def binary_to_target_arch(filename, target_arch, display_name=None):
+    """
+    Check that the given binary contains required architecture slice(s)
+    and convert the fat binary into thin one, if necessary.
+    """
+    if not display_name:
+        display_name = filename  # Same as input file
+    # Check the binary
+    is_fat, archs = get_binary_architectures(filename)
+    if is_fat:
+        if target_arch == 'universal2':
+            return  # Assume fat binary is universal2; nothing to do
+        else:
+            assert target_arch in archs, \
+                f"{display_name} does not contain slice for {target_arch}!"
+            # Convert to thin arch
+            logger.debug("Converting fat binary %s (%s) to thin binary (%s)",
+                         filename, display_name, target_arch)
+            convert_binary_to_thin_arch(filename, target_arch)
+    else:
+        assert target_arch != 'universal2', \
+            f"{display_name} is not a fat binary!"
+        assert target_arch in archs, \
+            f"{display_name} is incompatible with target arch " \
+            f"{target_arch} (has arch: {archs[0]})!"
+        return  # Nothing to do
+
+
+def remove_signature_from_binary(filename):
+    """
+    Remove the signature from all architecture slices of the given
+    binary file using the codesign utility.
+    """
+    logger.debug("Removing signature from file %r", filename)
+    cmd_args = ['codesign', '--remove', '--all-architectures', filename]
+    retcode, stdout, stderr = exec_command_all(*cmd_args)
+    if retcode != 0:
+        logger.warning("codesign command (%r) failed with error code %d!\n"
+                       "stdout: %r\n"
+                       "stderr: %r",
+                       cmd_args, retcode, stdout, stderr)
+        raise SystemError("codesign failure!")
+
+
+def sign_binary(filename, identity=None):
+    """
+    Sign the binary using codesign utility. If no identity is provided,
+    ad-hoc signing is performed.
+    """
+    if not identity:
+        identity = '-'  # ad-hoc signing
+
+    logger.debug("Signing file %r", filename)
+    cmd_args = ['codesign', '-s', identity, '--force', '--all-architectures',
```

Comment:
Should we specify `--options runtime` and `--timestamp` to ease notarization?
Another thing to consider, eventually, would be to add another option to allow passing an entitlements file (`--entitlements path/to/entitlements.plist`).

With all those parameters, one would be allowed to completely sign the app without having to post-process again all files after PyInstaller, which is the case for all code-signed softwares using it for now.

Reply:
Are these options relevant to all collected binaries, or only to the generated executable (or perhaps only to .app bundle)?

Reply:
To everything packaged into an app, and to the final app also.

Reply:
FTR here is the post-process I am using to sign a .app: https://github.com/nuxeo/nuxeo-drive/blob/9a5827710f113ba6ff61124adbbc7cb3b58b462a/tools/osx/deploy_ci_agent.sh#L108-L133 (put aside the extension stuff, it is not revelant here).

Reply:
Does hardened runtime require a real codesign identity? Trying to use it in combination with ad-hoc signing or with self-signed certificate results in error:

```
[2171] Error loading Python lib '/Users/rok/Development/pyi-macos11/dist/program/Python': dlopen: dlopen(/Users/rok/Development/pyi-macos11/dist/program/Python, 10): no suitable image found.  Did find:
	/Users/rok/Development/pyi-macos11/dist/program/Python: code signature in (/Users/rok/Development/pyi-macos11/dist/program/Python) not valid for use in process using Library Validation: mapped file has no Team ID and is not a platform binary (signed with custom identity or adhoc?)
```

Reply:
That's a good question, I guess ad-hoc signing is not compatible with hardened options (I just discovered ad-hoc signing with that PR). We could add a condition on the identity to enable or not hardened options.

Reply:
OK, we now enable hardened runtime if identity is provided. This leaves out signing with self-signed certificates, but I suppose in real-life scenarios, there's not much point to those anyway (we either have ad-hoc signing, or signing with proper identity). 

And if it turns out that we do want to decouple the options, we can always introduce additional option switch...

Reply:
I would push for `--entitlements` to being able to use new options for the next PyInstaller release :)

Reply:
Added in the new commit. Can you give it a try?

Reply:
Of course, I am testing it right now.

Reply:
It is taking more time than I hope. I am stuck on:
```
MyApp.app: code has no resources but signature indicates they must be present
```
I will try to isolate the thing to understand what is not good. I have custom data, maybe it does not help. Will keep you updated.

Reply:
In your tests, do those commands pass?
```shell
$ codesign --verbose=4 --deep --strict file.app
$ spctl --assess --verbose file.app
```

Reply:
Hmm, I think you need to sign the .app bundle - this PR explicitly doesn't do that, so you still need to do it yourself. But you probably don't have to do --deep signing, since individual binaries are already signed.

Reply:
I've now added an automatic attempt to sign the .app bundle as well. In contrast to library collection, we don't consider a failure here to be fatal, just in case some manual adjustments to bundle are needed.